ENDL module and unit test

arekfu opened the pull request:
I am starting to implement a module to read nuclear data in ENDL format. This pull request contains an embryonic module (based on endf.pyx) and the associated unit test.

arekfu committed:
introduce endl module

arekfu committed:
Add endl.rst to docs/pyapi

arekfu committed:
Remove np.import_array() from endl.py

arekfu committed:
Add test_endl.py

arekfu committed:
Make endl.py pass the unit test

arekfu committed:
Merge branch 'develop' of https://github.com/pyne/pyne into endl

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,239 @@
+from pyne import nucname
+
+import numpy as np
+import re
```
PEP8: Import ordering.  First there are Python standard library errors (re), then third party libraries (numpy), then current package (pyne)

scopatz commented on tests/test_endl.py:
```diff
@@ -0,0 +1,58 @@
+    try:
+        assert_equal(obs_hash, exp_hash)
+    except AssertionError:
+        raise AssertionError("epdl97_eedl_Pb hash check failed; please try redownloading the epdl97_eedl_Pb data file.")
```
PEP8: 80 character limit.

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,239 @@
+from warnings import warn
+from pyne.utils import QAWarning
+
+import pyne.rxdata as rx
```
I mistook this as a `rxname` because [our style guide says that rx stands for "reaction"](http://pyne.io/devsguide/style_guide.html#variable-name-conventions).  Would you mind just doing `from pyne import rxname`? Thanks!

arekfu replied:
You mean `from pyne import rxdata`, I guess? This line is also straight from ENDF, though.

scopatz replied:
OK.  It should probably be fixed there too.  (Though that isn't something you need to do.)

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,239 @@
+#        fh.seek(0)
+#        if opened_here:
+#            fh.close()
+#        return data
```
Remove commented out code.

scopatz replied:
Please.

arekfu replied:
I will soon uncomment and rework it.

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,239 @@
+
+
+
+    def _linlin(self, e_int, xs, low, high):
```
These functions don't use self at all and could / should be moved off of the class and then have unit tests.

scopatz replied:
It also looks like these might be copied from ENDF or somewhere.  Is that true?  If it is, then we should make these functions first class and put them somewhere else, perhaps in bins.py or their own module?

arekfu replied:
Yes, they are copied from ENDF. We should probably move them.

gonuke replied:
This is one of a few different interpolation functions.  Do any of the already included modules - I'm thinking numpy here - have interpolation routines already?  If not, these should perhaps live in some utility module somewhere.

scopatz replied:
> This is one of a few different interpolation functions. Do any of the
> already included modules - I'm thinking numpy here - have interpolation
> routines already? If not, these should perhaps live in some utility module
> somewhere.

NumPy does have interpolations functions but I think that the value here is
that they set the bounds correctly to match what is in the spec.  Some of
the implementation could possibly be cleaned up with that in mind.

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,239 @@
+#                            'particle_out': [],
+#                            'data': {},
+#                            'start': start,
+#                            }})
```
Remove commented out lines.

arekfu replied:
Already uncommented in future commit.

scopatz commented on tests/test_endl.py:
```diff
@@ -0,0 +1,58 @@
+    from StringIO import StringIO
+except ImportError:
+    from io import StringIO
+from math import e
```
No need for this import.  Just use `np.e`

scopatz commented on tests/test_endl.py:
```diff
@@ -0,0 +1,58 @@
+import warnings
+try:
+    from StringIO import StringIO
+except ImportError:
```
I am assuming that this is because of Python 2/3 compatibility.  In this case you should use something like the following.

``` python
import sys
if sys.version_info[0] > 2:
   from io import StringIO
else:
   from StringIO import StringIO
```

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,239 @@
+            line2 = fh.readline()
+
+            # EOF?
+            if not line2:
```
It is better to test `if  len(line2) != 0` than `if not ''`

arekfu replied:
Why? Just curious.

scopatz replied:
It is a more precise and explicit test.  What you care about here is that the list is zero, not that the the boolean happens to evaluate to false.  It is mostly for readability.

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,239 @@
+#!/usr/bin/env python
```
This module has no main function and should remove this line.

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,239 @@
+
+    def _read_headers(self):
+        opened_here = False
+        if isinstance(self.fh, basestring):
```
basestring is not Python 3 compatible. In other modules in pyne, this variable is defined as str in the event of Python 3.

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,239 @@
+        opened_here = False
+        if isinstance(self.fh, basestring):
+            fh = open(self.fh, 'r')
+            opened_here = True
```
Where is the file handle closed if we opened it?

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,239 @@
+
+            # skip to the end of the table
+            line = fh.readline()
+            while line and not END_OF_TABLE_RE.match(line):
```
`while len(line) > 0 ...`

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,239 @@
+
+
+
+class NotFound(Exception):
```
This class isn't used anywhere and probably should be removed.  Also, I am wary of new exceptions :).

scopatz commented on tests/test_endl.py:
```diff
@@ -0,0 +1,58 @@
+import nose
+from nose.tools import assert_equal
+
+def ignore_future_warnings(func):
```
What is the purpose of this function?  It isn't used.

scopatz commented on tests/test_endl.py:
```diff
@@ -0,0 +1,58 @@
+        try:
+            import urllib.request as urllib
+        except ImportError:
+            import urllib
```
This import should happen at the top of the module and should use the `sys.version_info` check.  See other test modules for examples.

scopatz commented on tests/test_endl.py:
```diff
@@ -0,0 +1,58 @@
+        except ImportError:
+            import urllib
+        urllib.urlretrieve("https://www-nds.iaea.org/epdl97/data/endl/eedl/za082000",
+                    "epdl97_eedl_Pb")
```
Indentation line up.

arekfu replied:
I'm a bit surprised that the `pep8` checker did not pick this up!

scopatz commented on tests/test_endl.py:
```diff
@@ -0,0 +1,58 @@
+def test_loadfile():
+    try:
+        assert(os.path.isfile('epdl97_eedl_Pb'))
+    except AssertionError:
```
This should be an if-statement rather than an assertion.

scopatz commented on tests/test_endl.py:
```diff
@@ -0,0 +1,58 @@
+    exp_hash = "502105669e0c0ad917301d219c649aaf"
+    try:
+        assert_equal(obs_hash, exp_hash)
+    except AssertionError:
```
No need to catch this.  I think you can pass in the message to `assert_equal()`

scopatz commented on tests/test_endl.py:
```diff
@@ -0,0 +1,58 @@
+    testlib = Library("epdl97_eedl_Pb")
+#    exp_nuclides = [10010000, 30060000, 40090000, 50110000, 30070000, 60000000, 50100000]
+#    obs_nuclides = list(map(int, testlib.structure.keys()))
+#    assert_array_equal(exp_nuclides, obs_nuclides)
```
Are these commented out for a reason?  Some quantitative tests would be nice.

arekfu replied:
They are commented out because, as things stand, `endl.Library` does not fill its internal structure dictionary (yet). The commented code is there to remind me to add the test as soon as the functionality is implemented.

scopatz commented:
Hi @arekfu!  Thanks for opening this.  I think ENDL support would be super valuable.  Also, thanks for keeping the initial code small.  I have done a pretty thorough review.  There are a couple of style issues which should be easy to fix.

scopatz commented:
Also, I didn't get to add this before, but I think there are some architectural issues that are important but not difficult to address.  It is good that we can have a discussion early.  I hope that this conversation helps!

arekfu commented:
Hi @scopatz, thank you for your review. Many of your remarks actually apply to endf.pyx and test_endf.py, since I copy-pasted and modified most of the code.

scopatz commented:
> Hi @scopatz, thank you for your review. Many of your remarks actually apply to endf.pyx and test_endf.py, since I copy-pasted and modified most of the code.

I am sure :).  I'll make a different issue to do a style review of that module as well.

arekfu committed:
Code review + PEP8 cleanliness

arekfu committed:
Fix import rxdata statement

scopatz commented:
Also, @arekfu - let us know when you'd like us to take a look again!

arekfu committed:
Populate the ENDL structure dictionary with table information

arekfu committed:
Do some quantitative tests on ENDL library parsing

arekfu commented:
Hi @scopatz, the commits above introduce parsing of the ENDL table headers. I have also beefed up the unit test. Can you please take a look at this? Sorry if this is taking some time, I am a bit busy right now.

arekfu committed:
Merge branch 'develop' of https://github.com/pyne/pyne into endl

scopatz commented:
Hi @arekfu - no worries on the delay :).  It looks like this is not Python 3 compliant.  I'll take a look.

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,264 @@
+
+        return {
+                'data_tuples': list()
+                }
```
Please put this one line and use a list literal.  ie:

``` python
return {'data_tuples': []}
```

Also, I am not sure something so simple needs its own function, but I am not opposed to it.

arekfu replied:
I need a factory function for `defaultdict`. It can be a `lambda`, though.

scopatz replied:
Sounds good to me.

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,264 @@
+    def structure_dict_entry():
+        """Static method to generate entries for the structure dict."""
+
+        return {
```
Please remove the newline.

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,264 @@
+
+    def _read_headers(self):
+        opened_here = False
+        if isinstance(self.fh, str):
```
The str here probably needs to be python 3 compatible.  Look in other modules for code that looks like 

``` python
if sys.version_info[0] > 2:
    basestring = str
```

and then test against basestring here.

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,264 @@
+class Library(rxdata.RxLib):
+    """A class for a file which contains multiple ENDL tables."""
+
+    DataTuple = namedtuple('DataTuple', ['yo', 'limits', 'x1', 'data'])
```
Probably best to move this to global scope rather than living on the library class.  Also nice use of namedtuple :)

scopatz commented on tests/test_endl.py:
```diff
@@ -0,0 +1,66 @@
+def test_loadfile():
+    if not os.path.isfile('epdl97_eedl_Pb'):
+        urllib.urlretrieve(
+                "https://www-nds.iaea.org/epdl97/data/endl/eedl/za082000",
```
This isn't a big deal but our style guide says to use single quotes for string literals (rather than double quotes) when it makes sense to do so. http://pyne.io/devsguide/style_guide.html#python-style-guide

scopatz commented:
Thanks @arekfu. This is getting very close.  I think you just need to fix the issue that is causing the Python 3 failure.  For reference you can see the failing test here: https://travis-ci.org/pyne/pyne/jobs/62099575#L2614

arekfu committed:
Suppress useless commented method

arekfu committed:
Python3 compatibility: should use isinstance(*, basestring)

arekfu committed:
Refactor small static method into a lambda

arekfu committed:
Move DataTuple out of Library class

arekfu committed:
Use single quotes for string literals

arekfu committed:
Use a list literal in defaultdict lambda factory

arekfu committed:
Merge remote branch 'origin/develop' into endl

arekfu committed:
Add docstring for endl.Library._read_headers

arekfu committed:
Add ENDL token parser to utils module

arekfu committed:
Fix position of end-of-table character in ENDL

arekfu committed:
Python3 compatibility fixes

arekfu committed:
Add leading underscore to private static method name

arekfu commented:
Yay! Python 3 works now.

arekfu committed:
Use None as placeholder for data yet to be read

arekfu committed:
Parse ENDL table as a numpy array

arekfu committed:
Add test file for ENDL module

arekfu committed:
PEP8 cleanliness

arekfu committed:
Extend ENDL unit test: parse some tables

arekfu committed:
Make x1 an int

The full ENDL specification actually says that this value is a float,
but for the purpose of electron-induced data it is always an integer (it
is used to indicate subshells). So let's make it an int and let us worry
tomorrow about tomorrow.

arekfu committed:
Remove data member from endl.DataTuple

Caching will be performed at the level of the data source.

arekfu committed:
Fix parsing and selection of ionization data

arekfu committed:
Beef up ENDL parsing unit tests

arekfu commented:
Another batch of updates:
- the Python 3 tests run fine;
- the module is able to do some serious parsing;
- new `fromendl_tok` function in the utils module.

Also, I gave up the idea of caching parsed data in the `Library` object because I realised that this is much better done at a higher level, by the to-be-written `ENDLDataSource` class.

The `Library` class seems more or less complete to me. However, I am not sure about the way I should handle double-differential data (such as angle-energy differential cross sections for photon emission in bremsstrahlung). For the moment I just return a big, 3-column `ndarray`. Perhaps it would make more sense to clearly expose the double-differential structure, but I'm not sure how to do this. How is this handled in the ENDF pipeline?

scopatz commented:
> Perhaps it would make more sense to clearly expose the double-differential structure, but I'm not sure how to do this. How is this handled in the ENDF pipeline?

I don't know... But it is probably a 2D numpy array.

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,341 @@
+
+DataTuple = namedtuple('DataTuple', ['yo', 'limits', 'x1'])
+
+nfields_rprop = {
```
PEP8 global constants should be in all upercase

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,341 @@
+
+        try:
+            pdp_dict = self.structure[nuc]['pin_rdesc_rprop']
+        except KeyError as e:
```
It is generally better to use `if key in ...` when you can rather than relying on exception handling.

scopatz replied:
Especially since you are just reraising the error.  I would rewrite this as:

``` python
if nuc in self.structure and 'pin_rdesc_rprop' in self.structure[nuc]:
    pdp_dict = self.structure[nuc]['pin_rdesc_rprop']
else:
    msg = ....
    raise KeyError(msg)
```

arekfu replied:
Guess where this code snippet comes from :)

I mildly disagree. I think that the `try`/`except` style is more [pythonic](https://docs.python.org/3.4/glossary.html#term-eafp). Interestingly, dictionary lookup is exactly the example that the [LBYL](https://docs.python.org/3.4/glossary.html#term-lbyl) entry puts forward.

scopatz replied:
So I always interpreted that to mean that 

``` python
pdp_dict = self.structure[nuc]['pin_rdesc_rprop'] 
```

on its on is more Pythonic (which it is), and if an error occurs you catch it somewhere farther up in the stack. Also the EAFP comparison is a little disingenuous since C exception handling syntax. However, if you are going to test for inclusion at the call site (though exceptions or conditionals), I think that there is an argument to be made for using conditionals based on their timings.  Consider the following:

``` python
scopatz@athenaie ~ $ def f(test):
`·.,¸,.·*¯`·.,¸,.·*¯     if test:
`·.,¸,.·*¯`·.,¸,.·*¯         pass
`·.,¸,.·*¯`·.,¸,.·*¯     else:
`·.,¸,.·*¯`·.,¸,.·*¯         pass
`·.,¸,.·*¯`·.,¸,.·*¯ 
scopatz@athenaie ~ $ def g(test):
`·.,¸,.·*¯`·.,¸,.·*¯     try:
`·.,¸,.·*¯`·.,¸,.·*¯         assert test
`·.,¸,.·*¯`·.,¸,.·*¯     except AssertionError:
`·.,¸,.·*¯`·.,¸,.·*¯         pass
`·.,¸,.·*¯`·.,¸,.·*¯ 
scopatz@athenaie ~ $ timeit "f(True)"
The slowest run took 130.28 times longer than the fastest. This could mean that an intermediate result is being cached.
10000000 loops, best of 3: 76.8 ns per loop
scopatz@athenaie ~ $ timeit "f(False)"
10000000 loops, best of 3: 75.8 ns per loop
scopatz@athenaie ~ $ timeit "g(True)"
10000000 loops, best of 3: 91.4 ns per loop
scopatz@athenaie ~ $ timeit "g(False)"
The slowest run took 59.51 times longer than the fastest. This could mean that an intermediate result is being cached.
1000000 loops, best of 3: 387 ns per loop
```

The False case for exceptions is _much_ slower.

arekfu replied:
> The False case for exceptions is much slower

That is true in most languages, I guess. But the point here is re-raising the exception with a more informative error message; so I would argue timing does not matter.

scopatz commented on pyne/endl.py:
```diff
@@ -0,0 +1,341 @@
+        def match_x1_yo_to_data_tuple(x, y, d):
+            return (x is None or d.x1 == x) and (y is None or d.yo == y)
+        try:
+            index, data_tuple = next((i, d) for i, d in enumerate(data_tuples)
```
Can't you just test if the length of `data_tuples` >0

arekfu replied:
No, I'm looking for the first data tuple that matches `x1`, `yo` or both, if they are not `None`. The `enumerate` is useless, however. I used it for the caching but then forgot to remove it.

scopatz replied:
But next() is only called once here... so it will either get the first element or nothing. Right?

arekfu replied:
That's right. I want the first element matching `x1` (if given) and `yo` (if given).

scopatz replied:
Ahh ok.

scopatz commented on tests/files_test_endl/testfile:
```diff
@@ -0,0 +1,2545 @@
+ 82000  9  0  0.00000+ 0 8912015 2  0.00000+ 0  0.00000+ 0  0.00000+ 0  
+ 7  0  0  0.00000+ 0  0.00000+ 0  0.00000+ 0  0.00000+ 0  0.00000+ 0
```
Where does this file come from and are you/we allowed to redistribute it?

arekfu replied:
I started from an EPDL data file (the one downloaded by the unit test), removed some tables and randomized the data in the remaining ones. I guess it's OK?

scopatz replied:
Yup that is OK, since it is randomized, thanks!

scopatz commented:
This is looking great @arekfu! Just a couple of minor issues to address.  Do you want a us to wait for the merge until after the double differential issue is addressed? Or would you prefer that to be in a separate pull request.

arekfu commented:
I'd like to fix these points first. I would also like to check if there is an easy solution for the interpolation of the double-differential data.

scopatz commented:
Hey @arekfu - please let us know when this is ready!

arekfu committed:
Use uppercase for endl.NFIELDS_RPROP global data member

arekfu commented:
OK, @scopatz, let's do this. I guess we can merge if you agree.

scopatz commented:
ooof I am so sorry about the delay in looking at this @arekfu. I have been on vacation ~_~.

scopatz commented:
Thanks a ton @arekfu!

scopatz commented:
And sorry again about the delay!
